add detailed error message

muescha opened the pull request:
add a more detailed error message if `git push` fails


example output:

```
Is everything correct? ([y]es / [n]o / [e]dit) y                                                                                                                        
Submitting…                                                                                                                                                             
There was an error submitting the pull request.                                                                                                                         
Please check the error message from the git command:                                                                                                                    
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
remote: error: GH007: Your push would publish a private email address.                                                                                                  
remote: You can make your email public or disable this protection by visiting:                                                                                          
remote: http://github.com/settings/emails                                                                                                                               
To https://github.com/muescha/homebrew-cask.git                                                                                                                         
 ! [remote rejected]       cask_repair_update-smartgit -> cask_repair_update-smartgit (push declined due to email privacy restrictions)                                 
error: failed to push some refs to 'https://github.com/muescha/homebrew-cask.git'                                                                                       
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
                                                                                                                                                                        
Please fix at first possible git problems.                                                                                                                              
Otherwise open a bug report on the repo for this script (https://github.com/vitorgalvao/tiny-scripts). 
```

vitorgalvao commented on cask-repair:
```diff
@@ -561,12 +561,23 @@ for cask in "${@}"; do
+    warning_message 'There was an error submitting the pull request.'
+    warning_message 'Please check the error message from the git command:'
+    divide
+    warning_message "$(cat "${submission_error_log}")"
```
[Useless use of `cat`](http://porkmail.org/era/unix/award.html). Also, make it an `echo` instead; so much yellow would be hard to read: `echo "$(< "${submission_error_log}")"`.

vitorgalvao commented on cask-repair:
```diff
@@ -561,12 +561,23 @@ for cask in "${@}"; do
+  fi
+
+  if [[ "${gitpush_result}" -ne 0 ]] && grep --quiet 'error:' "${submission_error_log}"; then
+    warning_message 'There was an error submitting the pull request.'
```
No repeated `warning_message`. Make it a single one instead: `warning_message 'There were errors while pushing:`.

vitorgalvao commented on cask-repair:
```diff
@@ -561,12 +561,23 @@ for cask in "${@}"; do
+  if [[ "${gitpush_result}" -ne 0 ]] && grep --quiet 'error:' "${submission_error_log}"; then
+    warning_message 'There was an error submitting the pull request.'
+    warning_message 'Please check the error message from the git command:'
+    divide
```
Remove. It doesn’t make sense to separate the message from the result.

vitorgalvao commented on cask-repair:
```diff
@@ -561,12 +561,23 @@ for cask in "${@}"; do
+    warning_message 'Please check the error message from the git command:'
+    divide
+    warning_message "$(cat "${submission_error_log}")"
+    divide
```
Remove this one as well.

vitorgalvao commented on cask-repair:
```diff
@@ -561,12 +561,23 @@ for cask in "${@}"; do
+    divide
+    warning_message "$(cat "${submission_error_log}")"
+    divide
+    abort 'Please fix at first possible git problems. \nOtherwise open a bug report on the repo for this script (https://github.com/vitorgalvao/tiny-scripts).'
```
Wording is weird. Make it instead `abort 'Please fix the errors and try again. If the issue persists, open a bug report on the repo for this script (https://github.com/vitorgalvao/tiny-scripts).'`

muescha replied:
i am not the best in working - fixed

muescha commented on cask-repair:
```diff
@@ -561,12 +561,23 @@ for cask in "${@}"; do
 
   git commit "${cask_file}" --message "${commit_message}" --quiet
   git push --force "${cask_repair_remote_name}" "${cask_branch}" --quiet 2> "${submission_error_log}"
+  gitpush_result="${?}"
```
i need the `gitpush_result` otherwise the `${?}` only hold the last result of previous command and i can not check the result in an second `if` in line 574

vitorgalvao replied:
Both times you’re saving the result of the command and then checking it right away. This variable is useless.

muescha replied:
not at all useless

the idea is to have 2 possible pull requests which can fail and only one check for it. thats why i set the `gitpush_result` 2 times and also the fill again the log with `submission_error_log`

if i check it with `${?}` in line 567 then the result is gone and i can not check the result 563 in line 574.

i like to check it also in the case when the first `if` is not true

muescha replied:
when i do it like this:

```
  git push --force "${cask_repair_remote_name}" "${cask_branch}" --quiet 2> "${submission_error_log}"

  if [[ "${?}" -ne 0 ]] && grep --quiet 'shallow update not allowed' "${submission_error_log}"; then
    echo 'Push failed due to shallow repo. Unshallowing…'
    HOMEBREW_NO_AUTO_UPDATE=1 brew tap --full "caskroom/$(current_tap)"
    git push --force "${cask_repair_remote_name}" "${cask_branch}" --quiet 2> "${submission_error_log}"
  fi

  if [[ "${?}" -ne 0 ]]; then 
```
then the second `if [[ "${?}"` not catch the error.

vitorgalvao replied:
That’s not what happens in my test:

```
if [[ 'a' == 'a' ]]; then # Just so it continues
  curl 'fake data' 2> /dev/null # This produces an error
fi

echo "${?}" # It correctly outputs a `6`, the result of the `curl` command.

muescha replied:
thats correct. for this case we don't need an variable.


this is when the first `if` is true. 

```
curl 'fake data' 2> /dev/null # This produces an error nr 1

if [[ "${?}" -ne 0 ]] && [[ 'b' == 'a' ]]; then # this is not true because b!=a
  curl 'fake data' 2> /dev/null # This produces an error nr 2 - but is never executed
fi

if [[ "${?}" -ne 0 ]]; then
  # this is never called because here is it always 0 and not the result from error 1
if
```

i need to check it if curl 1 makes an error but curl 2 never exected

vitorgalvao replied:
I see what you mean. I’d still rather deal with it a different way. Remove the `gitpush_result` and make it a nested `if`:

```
if [[ "${?}" -ne 0 ]]; then
  if grep --quiet 'shallow update not allowed' "${submission_error_log}"; then
    …
  else
    …
  fi
fi
```

This way we have a clearer section to deal with other errors in the future, instead of keeping overwriting the same variable.

muescha commented on cask-repair:
```diff
@@ -561,12 +561,23 @@ for cask in "${@}"; do
     HOMEBREW_NO_AUTO_UPDATE=1 brew tap --full "caskroom/$(current_tap)"
-    git push --force "${cask_repair_remote_name}" "${cask_branch}" --quiet
+    git push --force "${cask_repair_remote_name}" "${cask_branch}" --quiet 2> "${submission_error_log}"
+    gitpush_result = "${?}"
```
oh - wrong assignment - i will fix

muescha commented:
note: i removed the check for error in the log file because there are more errors possible (fatal and others)

muescha commented:
fixed to nested if

i am not shure should if i better give the error message `'There were errors while pushing:'` as parameter of `push_failure_message` (and renamed to `detailed_failure_message`) to be more generic?

and then call it like:
```
detailed_failure_message 'There were errors while pushing:' "$(< "${submission_error_log}")"
```

vitorgalvao commented on cask-repair:
```diff
@@ -561,20 +566,19 @@ for cask in "${@}"; do
+      echo 'Push failed due to shallow repo. Unshallowing…'
+      HOMEBREW_NO_AUTO_UPDATE=1 brew tap --full "caskroom/$(current_tap)"
+      git push --force "${cask_repair_remote_name}" "${cask_branch}" --quiet 2> "${submission_error_log}"
+      if [[ "${?}" -ne 0 ]]; then
```
Add an empty line before it and replace the block with `[[ "${?}" -ne 0 ]] && push_failure_message "$(< "${submission_error_log}")"`.

After that, I think we’re good to go!

muescha replied:
ok

vitorgalvao commented:
Looks good now. Thank you for the work and patience!
